### ISSUE TITLE
Version 2 reshuffle and refresh

### ISSUE THREAD

**Issue by mgeisler** [google/comprehensive-rust#1465]
@djmitche has been working on a major reshuffle and refresh of the course for a few months now: #1073.  The branch is getting to a state where we need to think about how to land this.

The branch is

- Adding new content
- Moving existing content
- Removing some existing content

From a read through, I think this is a significant improvement. It covers things in a much better order and it gives many new details!

There are some things I believe we need to address to merge it:

- How does this interfere with existing classes organized at Google? We have multiple overlapping classes right now, with the next open slot being Nov 22nd.
  - I feel we need to run this with real participants before we merge it. I would prefer an in-person class since those have more engagement in my experience.
  - We need to align merging this with the existing instructors.
- How does this interact with with translations?
  - #1463: I tried to summarize the impact there. The v2 branch touches around 500 of the ~2k messages in the course.
  - Because of the snapshot publishing system, it's should be okay to merge the branch.
  - Overall, I'm impressed how much of the existing text has been preserved!
- How does this interact with existing pages?
  - #489: We don't have a good understanding of the number of 404 hits we have on the course.
  - #488: Nor do we know how ofte the redirection pages are used.
  - I'm afraid that we won't notice problems here if we just land the branch without data on this.
- Size and style of the new slides
  - #1464: I tried to write down my thoughts there.
  - I'm worried about being able to reign this in if we suddenly let several of the slides become 50% larger. When I've talked with people about this in the past, I've sometimes been met with a "well it fits on my screen"-style reply. Which doesn't solve the actual problem I observe in class and leaves me to pick up the pieces.

**Comment by mgeisler** [comment]
cc @randomPoison, @fw-immunant, @henrif75, @sakex, @fbornhofen, @deavid, etc who are teaching classes, have reviewed the branch, etc...

**Comment by djmitche** [comment]
I would add to this

 * Suggested changes in the ordering of segments in days 2-3
   * https://github.com/google/comprehensive-rust/pull/1073#issuecomment-1804734058: Moving "Memory Management" a little earlier so that "Borrowing" and "Slices and Lifetimes" can be adjacent.
   * This is relatively simple, but requires updating exercises to ensure they don't require any un-covered content.
 * Total time for the course increases by something like 30%
   * Partly due to more exercises, partly due to filling in some content gaps in v1
   * Do we cut content, or make this a 4-day course (https://github.com/google/comprehensive-rust/pull/1073#issuecomment-1805023928)?

I'm concerned that this issue encompasses a _substantial_ amount of additional work -- likely well into 2024.

**Comment by mgeisler** [comment]
> * Total time for the course increases by something like 30%
>   
>   * Partly due to more exercises, partly due to filling in some content gaps in v1
>   * Do we cut content, or make this a 4-day course ([Comprehensive Rust v2 #1073 (comment)](https://github.com/google/comprehensive-rust/pull/1073#issuecomment-1805023928))?

I think this is the biggest blocker for merging this right now: it'll make it awkward for us to teach it at Google if we know that we'll run out of time on every day. So we should figure out ASAP if we'll cut content or go with a 4-day class.

Note that there is no rule that says that we cannot make this a 4-day course! Students have been repeatedly asking for more exercises and this rewrite gives them that :slightly_smiling_face: There are different ways in which we could structure a 4-day class:

- **Two 2-day classes:**
  - Pros: people can sign up for the first half one month and then take the second half at a later point. Instructors also have more flexibility this way.
  - Cons: people who only take the first two days won't really be fully equiped to go back to their teams and be productive with Rust.

- **One 3-day class plus a 1-day class:** This could be structured like a "Rust Fundamentals" plus a "Advanced Rust" class.
  - Pros: People will hopefully be productive with just the 3-day class.
  - Cons: ?

  A big question is how we can make such a cut in a useful way? Candidates for an Advanced Rust class:
    - Unsafe Rust
    - Interoperability with C++ and Java? This would steal content from the Android class.
    - Error handling?
    - Smart pointers?
  It's not really clear to me what we would put on this Day 4 since we want everybody to know about things like error handling as well as `Box` and `Rc`, etc. Unsafe Rust is the part that fits best into an Advanced Rust class in my opinion.

- **One 4-day class**
  - Pros: Simple format, we can fit our content into it!
  - Cons: The class is now firmly in the "whole week" territory. We normally teach the class spread over two weeks, so this should be okay. We also started with a 4-day class until we split off the Android class.

While I like the flexibility of two 2-day classes, I lean towards going with a single 4-day class because of how it allows us to make a class where we can be confident will make people productive in Rust.

**Comment by djmitche** [comment]
If we choose 3 + 1, I would suggest that the "Advanced Rust" course leaning more toward more advanced idioms and best-practices that aren't strictly part of the language. It's sort of an artificial division, but it means you have _something_ after 3 days. I think we should continue to recommend all four days, just maybe not taken consecutively. My thinking for that last day is something like:

 * Iterators (30m)
 * Testing (55m)
 * Error Handling (35m)
 * Unsafe (40m)

I think Smart Pointers is actually important to the narrative from general memory-management to move semantics and borrowing. But even adding some extra time for harder exercises, that list is only 3.5 hours - not a full day by any stretch.

So, maybe the four days is the better idea?

**Comment by djmitche** [comment]
I like that pitch :)

**Comment by djmitche** [comment]
OK, I've updated #1073 appropriately. The course is now 4 days, and I've made the suggested reorganizations of content. I have _not_ addressed the size and style, interactions with translations, or interactions with existing pages (I made some manual effort to update book.toml, but may have missed things).

I think we are at a spot where the branch can be merged and the remainder of the work done on `main`.
